Correctly parse errors on Windows

shivjm opened the pull request:
Errors from hledger weren’t recognized as such using Emacs 30 from Git (2023-05-03) with flycheck-hledger from Git (87ed9d4), hledger 1.30, and flycheck 33snapshot on Windows 10. Instead, I kept getting the errors as ‘Suspicious state’ messages in the minibuffer. Some digging showed the following problems:

1. The patterns expected `hledger:` but the output contained `hledger.exe:`.
2. The patterns forbade colons in filenames but, on Windows, these will typically contain colons.
3. A literal `\n` in a pattern wouldn’t match the line endings (I suppose this is because the output contained `\r\n`).

This PR fixes all of those issues and, as far as I can tell, correctly parses all the messages in [hledger’s error tests](https://github.com/simonmichael/hledger/tree/f4508e73d3dc3709bd403487dee56f8cb24eaa5b/hledger/test/errors). I unfortunately can’t test on a Linux machine at present; I don’t anticipate issues, but it’s always possible the line endings would be matched differently there, in which case perhaps the patterns would need something like `(or line-end "\n")`.

DamienCassou commented on flycheck-hledger.el:
```diff
@@ -77,39 +77,39 @@ https://hledger.org/hledger.html#check."
-    bol "hledger: Error: balance assertion: " (file-name (minimal-match (one-or-more (not ":")))) ":" line ":" column "\n"
-    (message (one-or-more bol (zero-or-more nonl) "\n")))
+    bol "hledger" (optional ".exe") ": Error: balance assertion: " (file-name) ":" line ":" column line-end
+    (message (one-or-more "\n" (zero-or-more nonl) line-end)))
```
what about

```emacs
(message (one-or-more anychar))
```

? If all tests still pass, I'm happy.

DamienCassou reviewed (changes requested):
Thank you very much for your work. I left a few comments.

I also just pushed a commit adding unit tests. Would you mind rebasing your code on top of the latest commit and run the tests to make sure everything is still passing?

Also, would you mind adding tests to check the changes? I hope you  just have to add a few `defconst` and update `flycheck-hledger-test-error-symbols` to reference them in `flycheck-hledger-test.el`.

shivjm commented:
Switching to `anychar` alone was fine. I added a test with an hledger message that failed, though:
```emacs-lisp
    :output "hledger.exe: Error: ./file.ledger:2:
  | 2022-01-01
2 |     (a)               1
  |      ^

Strict account checking is enabled, and
account \"a\" has not been declared.
Consider adding an account directive. Examples:

account a
account a    ; type:A  ; (L,E,R,X,C,V)

"
```
The existing patterns couldn’t handle the line number being on the second line of the excerpt and still parse the message correctly. I fixed this by adding stricter matching for excerpts. All the tests are now passing. (And I’m reasonably confident that if this one Windows case is parsed correctly they should all be.)

DamienCassou commented on flycheck-hledger.el:
```diff
@@ -77,39 +77,39 @@ https://hledger.org/hledger.html#check."
    (error
-    bol "hledger: Error: balance assertion: " (file-name (minimal-match (one-or-more (not ":")))) ":" line ":" column "\n"
-    (message (one-or-more bol (zero-or-more nonl) "\n")))
+    bol "hledger" (optional ".exe") ": Error: balance assertion: " (file-name) ":" line ":" column line-end
```
it would make sense that the commit introducing this fix also introduces the test for this fix. This way the commit is atomic and can be merged (and potentially later reverted) separately. An atomic commit is also easier to understand because it contains everything that makes sense together and nothing more.

DamienCassou commented on tests/flycheck-hledger-test.el:
```diff
@@ -107,6 +127,7 @@ The matched text is checked against expected values in ERROR-OBJECT."
   "Return non-nil if and only if ERROR-PATTERN matches the current buffer.
 The matched text is checked against expected values in ERROR-OBJECT."
   (goto-char (point-min))
+  (message "pattern: %S, text: %S" (car error-pattern) (buffer-string))
```
I don't think we want that

DamienCassou commented on tests/flycheck-hledger-test.el:
```diff
@@ -71,22 +71,42 @@ The error message.
+    flycheck-hledger-test-balance-error-windows))
 
 (ert-deftest flycheck-hledger-test-error-patterns ()
-  (let* ((error-patterns (flycheck-checker-get 'hledger 'error-patterns)))
```
this seems to be a bug

DamienCassou commented on tests/flycheck-hledger-test.el:
```diff
@@ -71,22 +71,42 @@ The error message.
+  '(
+    :expected-file "./file.ledger"
+    :expected-line "2"
+    :expected-message "Strict account checking is enabled, and\naccount \"a\" has not been declared.\nConsider adding an account directive. Examples:\n\naccount a\naccount a    ; type:A  ; (L,E,R,X,C,V)\n\n"
```
could this message be as simple as the ones of other error objects? This error object is only meant to test the behavior on Windows. Feel free to add another error object for a different issue you are fixing (which should probably go in a separate and atomic commit)

shivjm replied:
To clarify: since the other tests don’t use real messages, should I take a real error message from hledger and simplify it to match? (As mentioned below, I’m really having trouble figuring out why one pattern matched and the other didn’t, because the strings themselves look identical to the Linux ones rather than having differing line endings, so I’m not sure what to edit to match them.)

DamienCassou replied:
What I meant was to make as many error objects as bugs you are fixing instead of checking all bug fixes with just one error object.

DamienCassou commented on tests/flycheck-hledger-test.el:
```diff
@@ -71,22 +71,42 @@ The error message.
 The error message.
 "))
 
+(defconst flycheck-hledger-test-balance-error-windows
```
this error object should probably use `\r\n` to check the fixes you wrote

DamienCassou commented on flycheck-hledger.el:
```diff
@@ -77,39 +77,39 @@ https://hledger.org/hledger.html#check."
-    bol "hledger: Error: balance assertion: " (file-name (minimal-match (one-or-more (not ":")))) ":" line ":" column "\n"
-    (message (one-or-more bol (zero-or-more nonl) "\n")))
+    bol "hledger" (optional ".exe") ": Error: balance assertion: " (file-name) ":" line ":" column line-end
+    (message (one-or-more "\n" (zero-or-more nonl) line-end)))
```
it feels wrong to me that the `\n` character from the first line is now part of the message. Can we exclude it with, e.g.,


```emacs
"\n"
(message (one-or-more (zero-or-more nonl) line-end))
```

?

DamienCassou replied:
I've now seen that this is done in the final commit. Is there a way to organize the commits so they make sense individually?

DamienCassou commented on flycheck-hledger.el:
```diff
@@ -77,39 +77,39 @@ https://hledger.org/hledger.html#check."
-    bol "hledger: Error: " (file-name (minimal-match (one-or-more (not ":")))) ":" line "-" end-line "\n"
-    (message (one-or-more bol (zero-or-more nonl) "\n")))
+    bol "hledger" (optional ".exe") ": Error: " (file-name) ":" line "-" end-line line-end
+    (message (one-or-more anychar)))
```
this also makes the end of the previous line part of the message

DamienCassou commented on flycheck-hledger.el:
```diff
@@ -77,39 +77,39 @@ https://hledger.org/hledger.html#check."
-    bol "hledger: Error: balance assertion: " (file-name (minimal-match (one-or-more (not ":")))) ":" line ":" column "\n"
-    (message (one-or-more bol (zero-or-more nonl) "\n")))
+    bol "hledger" (optional ".exe") ": Error: balance assertion: " (file-name) ":" line ":" column line-end
+    (message (one-or-more "\n" (zero-or-more nonl) line-end)))
```
I'm surprised that this works on Windows because the end of line is supposed to consists of 2 characters `\r\n` but only `\n` is read here

shivjm replied:
I’m not sure why either. The behaviour here around newlines doesn’t make a lot of sense to me.

DamienCassou replied:
from what I understand `line-end` matches the empty string at the end of the line. This means that there is still the end of line character to match after `line-end` has matched.

This doesn't explain why the code doesn't need to match `\r` though... If it works for you I guess it's good enough.

DamienCassou reviewed (changes requested):
Thank you very much for this great work. I left a few comments, mostly about git history. Also, it seems the code can't be parsed and contains some leftovers. You should be able to test the quality of each commit by running `make check` at the root of the project.

shivjm commented:
Sorry, you’re definitely right that the commits have got a bit garbled. Let me see why that is.

shivjm committed:
Correctly parse errors on Windows

shivjm committed:
Correctly parse excerpts with line numbers anywhere in them

shivjm commented:
I went back to the drawing board and investigated what exactly caused the errors on Windows. Rather than line endings, it seems it was only the filenames with drive letters and colons, which I wasn’t catching in my tests because I used relative paths.

I believe this updated PR should do the trick. I’ve structured it as a pair of atomic commits: one for basic error parsing and one for parsing excerpts where the line number isn’t on the first line.

DamienCassou reviewed (changes requested):
great work again. Thanks for taking the time to investigate further. My only question now is about the second commit: it only contains new tests but it's message is "Correctly parse excerpts with line numbers anywhere in them". Is there something wrong?

shivjm commented:
Thanks for looking at it. That’s strange about the second commit… I remember splitting up the commits so that the changes to the patterns for the excerpts were in the second commit, but I must have accidentally reverted my changes to the changes along the way. Sorry about that! The second commit should now have the excerpt-related changes that I accidentally put in the first one.

DamienCassou reviewed (approved):
Awesome, thank you very much for your effort!

shivjm commented:
Thanks for your patience with all the odd Git shenanigans!
